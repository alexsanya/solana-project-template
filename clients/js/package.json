{
  "name": "svm-program-e2e",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "build-bpf": "cargo build-bpf --manifest-path ../Cargo.toml",
    "test": "COPYFILE_DISABLE=1 jest --forceExit"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@jest/globals": "^29.7.0",
    "@solana/web3.js": "^1.87.6",
    "jest": "^29.7.0",
    "ts-jest": "^29.1.1"
  },
  "solana": {
    "wallet": "../program.json",
    "output": "../target/sbf-solana-solana/release/solana_program.so"
  },
  "dependencies": {
    "crypto": "^1.0.1",
    "merkletreejs": "^0.5.1"
  }
}
