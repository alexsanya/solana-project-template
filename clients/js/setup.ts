process.env['RPC_PORT'] = process.env['RPC_PORT'] || '8899';
const RPC_PORT = process.env['RPC_PORT'];

module.exports = async function() {
  // Wait for RPC to be ready
  await waitForRpc();
}

async function waitForRpc(): Promise<boolean> {
  while (true) {
    if (await (validateRpc())) {
      console.log("RPC validation success");
      return true;
    }
  }
}

async function validateRpc(): Promise<boolean> {
  const data = await fetch(`http://127.0.0.1:${RPC_PORT}`, {
    method: 'post',
    body: JSON.stringify({
      jsonrpc: "2.0",
      id: "1",
      method: "getHealth"
    }),
    headers: { 'Content-Type': 'application/json' },
  })
  .then((res) => res.json())
  .catch(() => null)

  return data?.result === 'ok';
}