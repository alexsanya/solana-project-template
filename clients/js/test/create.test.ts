import {describe, test, expect} from '@jest/globals';
import { Connection, Transaction, TransactionInstruction, PublicKey, Keypair, sendAndConfirmTransaction, SystemProgram, SYSVAR_RENT_PUBKEY } from '@solana/web3.js';
import sk from './program.json';
const crypto = require('crypto');

const sha256 = (data: any) => crypto.createHash('sha256').update(data).digest();

describe('Merkle tree program', () => {
  const programId = new PublicKey("TREEZwpvqQN6HVAAPjqhJAr8BuoGhXSx34jm9YV5DPB");
  const wallet = Keypair.fromSecretKey(Uint8Array.from(sk));
  const port = process.env['RPC_PORT'];
  const connection = new Connection(`http://127.0.0.1:${port}`, 'confirmed');
  const SEED = 'tree';

  test("Insert leaf", async () => {
    console.log("Program ID:", programId.toBase58());

    const [pda, bump] = await PublicKey.findProgramAddressSync(
      [Buffer.from(SEED), wallet.publicKey.toBuffer()], // seeds
      programId
    );

    console.log("Derived PDA:", pda.toBase58());
    console.log("Bump:", bump);

    //create account
    const initializeTreeInstruction = new TransactionInstruction({
      keys: [
        { pubkey: wallet.publicKey, isSigner: true, isWritable: true },
        { pubkey: pda, isSigner: false, isWritable: true },
        { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
        { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false }
      ],
      programId: new PublicKey(programId),
      data: Buffer.from([0, 3])
    });

    console.log("Tree account sent to instruction ", pda.toBase58());

    const leaf = sha256(Buffer.from('LeafA'));
    const addLeafInstruction = new TransactionInstruction({
      keys: [
        { pubkey: wallet.publicKey, isSigner: true, isWritable: true },
        { pubkey: pda, isSigner: false, isWritable: true },
      ],
      programId: new PublicKey(programId),
      data: Buffer.concat([Buffer.from([1]), leaf])
    });

    const transaction = new Transaction()
      .add(initializeTreeInstruction)
      .add(addLeafInstruction)

    const sig = await sendAndConfirmTransaction(connection, transaction, [wallet]);

    const tx = await connection.getParsedTransaction(sig);
    const logs = tx?.meta?.logMessages || [];
    const selectedEvents = logs.filter(line => ['event:CreateTree', 'event:LeafInserted'].some(event => line.includes(event)));
    expect(selectedEvents).toMatchSnapshot();
  });
});
