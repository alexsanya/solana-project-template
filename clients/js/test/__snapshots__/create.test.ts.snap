// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merkle tree program Insert leaf 1`] = `
[
  "Program log: event:CreateTree address:HuYk7PA3TBzDsgdRsxpKhxzEcB1L9o3dHNNFcaYNEfs1",
  "Program log: event:LeafInserted NewRoot: 8dd037fbec59096df37b2b042da35ba594c511cf1d760a9b98a99be1946c3d24",
]
`;
